arroyo/types.py

```diff
@@ -85,6 +79,76 @@ def __setstate__(self, state: dict[str, int | datetime]) -> None:
             object.__setattr__(self, slot, value)
 
 
+class BasePayload(Generic[TPayload]):
+    @property
+    def payload(self) -> TPayload:
+        pass
+
+    @property
+    def committable(self) -> Mapping[Partition, Position]:
+        pass
+
+
+@dataclass(unsafe_hash=True)
+class Payload(BasePayload[TPayload]):
+    """
+    Any other payload that may not map 1:1 to a single message from a
+    consumer. May represent a batch spanning many partitions.
+    """
+
+    __slots__ = ["__payload", "__committable"]
+    __payload: TPayload
+    __committable: Mapping[Partition, Position]
+
+    def __init__(
+        self, payload: TPayload, committable: Mapping[Partition, Position]
+    ) -> None:
+        self.__payload = payload
+        self.__committable = committable
+
+    @property
+    def payload(self) -> TPayload:
+        return self.__payload
+
+    @property
+    def committable(self) -> Mapping[Partition, Position]:
+        return self.__committable
+
+
+@dataclass(unsafe_hash=True)
+class BrokerPayload(BasePayload[TPayload]):
+    """
+    A payload received from the consumer or producer after it is done producing.
+    Partition, offset, and timestamp values are present.
+    """
+
+    __slots__ = ["__payload", "partition", "offset", "timestamp"]
+    __payload: TPayload
+    partition: Partition
+    offset: int
+    timestamp: datetime
+
+    def __init__(
+        self, payload: TPayload, partition: Partition, offset: int, timestamp: datetime
+    ):
+        self.__payload = payload
+        self.partition = partition
+        self.offset = offset
+        self.timestamp = timestamp
+
+    @property
+    def payload(self) -> TPayload:
+        return self.__payload
+
+    @property
+    def committable(self) -> Mapping[Partition, Position]:
+        return {self.partition: Position(self.next_offset, self.timestamp)}
```

Comment:
Is there in the standard library any data structure that implements `Mapping` and that is cheaper to instantiate than a dictionary ? This happens thousands times per second on a consumer so even a minimal improvement here can have a meaningful impact.